### PR TITLE
Main namelist debug

### PR DESCRIPTION
**Pull request by daveh150**
For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>

## PR checklist
- [ ] Short (1 sentence) summary of your PR: 
    Provide detailed output when namelist has wrong variables
- [ ] Developer(s): 
    David Hebert
- [ ] Suggest PR reviewers from list in the column to the right.
- [ ] Please copy the PR test results link or provide a summary of testing completed below.
Quick suite BFB test on GAFFNEY shows:
21 measured results of 21 total results
21 of 21 tests PASSED
0 of 21 tests PENDING
0 of 21 tests MISSING data
0 of 21 tests FAILED

- How much do the PR code changes differ from the unmodified code? 
    - [ x] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ x] Yes
    - [ ] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [x ] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] Yes
    - [x ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ x] No 
- [ x] Please provide any additional information or relevant details below:
If you have a namelist variable incorrect (say CICE was updated with different namelist variable name), the error just tells you which namelist. This enhancement adds method to print the actual line in namelist. 

Example: CICE6 changed 'sst_data_type' to 'ocn_data_type'. If you didn't change it new code will now print:

 (abort_ice)ABORTED:
 (abort_ice) called from ice_init.F90
 (abort_ice) line number          579
 (abort_ice) error =
 (input_data)ERROR: reading namelist    sst_data_type   = 'gofs_daily'

### CODE REVIEW

**Comment by phil-blain** [comment]
Hi David, I think this could be really useful. It's unfortunate that we have to repeat the code in several files, tough (but that would be a subject for another day!)

I noticed that the second commit is simply a fixup of the first one. You could use an [interactive rebase](https://git-rebase.io) to clean that up if you want.

I also noticed that you used a different email address for the two commits...

**Comment by apcraig** [comment]
@daveh150.  This is a nice feature to have.  Knew we weren't handling this case well, but wasn't sure how to work around that.  Nice update.

**Comment by eclare108213** [comment]
I agree.  Does it make sense to pull all these bits of identical code into subroutine in CICE_Finalize (or somewhere)?

**Comment by daveh150** [comment]
Thank you Phillipe! I have one more update, I missed a namelist in cicecore/shared/ice_init_column.F90. Do I do an interactive rebase first, then push to my fork?

 

Thanks!

David

**Comment by phil-blain** [comment]
Yes, though you would have to push with `--force` since you would be changing history.

**Comment by daveh150** [comment]
I pushed the commit to cicecore/shared/ice_init_column.F90. I also did the interactive rebase Phillipe suggested. I did not need to use --force, but it seemed to work. :-)

File: cicecore/cicedynB/analysis/ice_history_drag.F90 [hunk 1]
```diff
@@ -89,16 +91,30 @@ subroutine init_hist_drag_2D
...
+
       endif
       call release_fileunit(nu_nml)
 
       call broadcast_scalar(nml_error, master_task)
+      call broadcast_scalar(tmpstr2,   master_task)
       if (nml_error /= 0) then
          close (nu_nml)
-         call abort_ice(subname//'ERROR: reading icefields_drag_nml')
+         call abort_ice(subname//'ERROR: reading icefields_drag_nml' // &
```
Comment (apcraig) [inline cicecore/cicedynB/analysis/ice_history_drag.F90:115]: Need a space at the end of the string like other calls.

File: cicecore/cicedynB/analysis/ice_history_fsd.F90 [hunk 2]
```diff
@@ -105,15 +107,28 @@ subroutine init_hist_fsd_2D
...
+         endif
       endif
       call release_fileunit(nu_nml)
 
       call broadcast_scalar(nml_error, master_task)
+      call broadcast_scalar(tmpstr2,   master_task)
       if (nml_error /= 0) then
          close (nu_nml)
-         call abort_ice(subname//'ERROR: reading icefields_fsd_nml')
+         call abort_ice(subname//'ERROR: reading icefields_fsd_nml'// &
```
Comment (apcraig) [inline cicecore/cicedynB/analysis/ice_history_fsd.F90:129]: need a space at the end of the ERROR string.

File: cicecore/cicedynB/analysis/ice_history_mechred.F90 [hunk 3]
```diff
@@ -111,15 +113,30 @@ subroutine init_hist_mechred_2D
...
+
       endif
       call release_fileunit(nu_nml)
 
       call broadcast_scalar(nml_error, master_task)
+      call broadcast_scalar(tmpstr2,   master_task)
       if (nml_error /= 0) then
          close (nu_nml)
-         call abort_ice(subname//'ERROR: reading icefields_mechred_nml')
+         call abort_ice(subname//'ERROR: reading icefields_mechred_nml' // &
```
Comment (apcraig) [inline cicecore/cicedynB/analysis/ice_history_mechred.F90:137]: need a space at the end of the ERROR string.

File: cicecore/cicedynB/general/ice_init.F90 [hunk 4]
```diff
@@ -563,39 +563,59 @@ subroutine input_data
...
+            read(nu_nml,fmt='(A)') tmpstr2 
+         endif
+      endif  ! if my_task == master_task
+
       call broadcast_scalar(nml_error, master_task)
+      call broadcast_scalar(tmpstr2,   master_task)
       if (nml_error /= 0) then
-         call abort_ice(subname//'ERROR: reading namelist', &
-            file=__FILE__, line=__LINE__)
+         call abort_ice(subname//'ERROR: reading namelist' // &
```
Comment (apcraig) [inline cicecore/cicedynB/general/ice_init.F90:616]: Need a space at the end of the ERROR string.

**Review by apcraig — CHANGES_REQUESTED** [review]
A few comments.  Since we're here, how about if we unify the implementation across different files to

         do while (nml_error > 0)
            read(nu_nml, nml=icefields_nml,iostat=nml_error)
            if (nml_error /= 0) exit
            ...
         end do

         if (nml_error /= 0) then
            backspace(nu_nml) 
            read(nu_nml,fmt='(A)') tmpstr2 
            call abort_ice(subname//'ERROR: reading namelist' // &
              trim(tmpstr2), &
              file=__FILE__, line=__LINE__)
         endif
         close(nu_nml)

**Comment by apcraig** [comment]
Sorry, an old comment got included in my review that wasn't supposed to be there.  I have removed it.

**Comment by daveh150** [comment]
I added the needed spaces and colon for output messages.

**Review by apcraig — APPROVED** [review]

**Review by eclare108213 — APPROVED** [review]

**Pull request merged**